clients/da-vinci-client/src/main/java/com/linkedin/davinci/notifier/DaVinciPushStatusUpdateTask.java

```diff
@@ -57,75 +63,149 @@ public void batchPushEndSignalSent() {
-          version.getStoreName(),
-          version.getNumber(),
-          ExecutionStatus.STARTED,
-          getTrackedPartitions());
+      sendPushStatus(ExecutionStatus.STARTED, Optional.empty());
       batchPushStartSignalSent();
-    } else if (!isBatchPushEndSignalSent() && areAllPartitionsOnSameTerminalStatus(ExecutionStatus.COMPLETED)
+    } else if (!isBatchPushEndSignalSent()
+        && areAllPartitionsOnSameTerminalStatus(ExecutionStatus.COMPLETED, Optional.empty())
         && areAllPartitionFuturesCompletedSuccessfully.get()) {
-      pushStatusStoreWriter.writeVersionLevelPushStatus(
-          version.getStoreName(),
-          version.getNumber(),
-          ExecutionStatus.COMPLETED,
-          getTrackedPartitions());
+      sendPushStatus(ExecutionStatus.COMPLETED, Optional.empty());
       batchPushEndSignalSent();
-      // Shutdown the scheduler after sending the final status
-      shutdown();
-    } else if (!isBatchPushEndSignalSent() && isAnyPartitionOnErrorStatus()) {
-      pushStatusStoreWriter.writeVersionLevelPushStatus(
-          version.getStoreName(),
-          version.getNumber(),
-          ExecutionStatus.ERROR,
-          getTrackedPartitions());
+      batchPushPartitionStatus.clear();
+    } else if (!isBatchPushEndSignalSent() && isAnyPartitionOnErrorStatus(Optional.empty())) {
+      sendPushStatus(ExecutionStatus.ERROR, Optional.empty());
       batchPushEndSignalSent();
-      // Shutdown the scheduler after sending the final status
-      shutdown();
+      batchPushPartitionStatus.clear();
     }
   }
 
-  /**
-   * Get the partition id set that is being tracked
-   */
-  public Set<Integer> getTrackedPartitions() {
-    return partitionStatus.keySet();
+  private void handleIncrementalPushStatus() {
+    incrementalPushVersionToStatus.forEach((incrementalPushVersion, incrementalPushStatus) -> {
+      if (!incrementalPushStatus.startSignalSent) {
+        sendPushStatus(ExecutionStatus.START_OF_INCREMENTAL_PUSH_RECEIVED, Optional.of(incrementalPushVersion));
+        incrementalPushStatus.startSignalSent = true;
+      } else if (!incrementalPushStatus.endSignalSent && areAllPartitionsOnSameTerminalStatus(
+          ExecutionStatus.END_OF_INCREMENTAL_PUSH_RECEIVED,
+          Optional.of(incrementalPushVersion))) {
+        sendPushStatus(ExecutionStatus.END_OF_INCREMENTAL_PUSH_RECEIVED, Optional.of(incrementalPushVersion));
+        incrementalPushStatus.endSignalSent = true;
+        // Clean up the status map for incremental push versions in case there are too many of them and cause OOM
+        incrementalPushStatus.partitionStatus.clear();
+      }
+    });
   }
 
-  public boolean areAllPartitionsOnSameTerminalStatus(ExecutionStatus status) {
-    if (partitionStatus.isEmpty()) {
-      return false;
+  private void sendPushStatus(ExecutionStatus status, Optional<String> incrementalPushVersion) {
+    pushStatusStoreWriter.writeVersionLevelPushStatus(
```

Comment:
Can you add a logging here for producing the push status message for easier debugging purpose?

Reply:
I just realize there is already a logging inside "PushStatusStoreWriter#writeVersionLevelPushStatus" before sending; I added an additional log in the callback when the produce is completed.

Reply:
I see. That also works.